wire-moshi-support/src/main/java/com/squareup/wire/WireJsonAdapterFactory.java

```diff
@@ -0,0 +1,141 @@
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.squareup.moshi.Moshi;
+import com.squareup.moshi.Types;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import okio.ByteString;
+
+/**
+ * A {@link JsonAdapter.Factory} that allows Wire messages to be serialized and deserialized using
+ * the Moshi Json library.
+ *
+ * <pre>
+ * Moshi moshi = new Moshi.Builder()
+ *   .add(new WireJsonAdapterFactory())
+ *   .build();
+ * </pre>
+ *
+ * The resulting {@link Moshi} instance will be able to serialize and deserialize any Wire {@link
+ * Message} type, including extensions and unknown field values. The JSON encoding is intended to be
+ * compatible with the
+ * <a href="https://code.google.com/p/protobuf-java-format/">protobuf-java-format</a>
+ * library. Note that version 1.2 of that API has a
+ * <a href="https://code.google.com/p/protobuf-java-format/issues/detail?id=47">bug</a>
+ * in the way it serializes unknown fields, so we use our own approach for this case.
+ */
+public final class WireJsonAdapterFactory implements JsonAdapter.Factory {
+  static final JsonAdapter<ByteString> BYTE_STRING_JSON_ADAPTER = new JsonAdapter<ByteString>() {
+    @Override public void toJson(JsonWriter out, ByteString byteString) throws IOException {
+      out.value(byteString.base64());
+    }
+
+    @Override public ByteString fromJson(JsonReader in) throws IOException {
+      return ByteString.decodeBase64(in.nextString());
+    }
+  }.nullSafe();
+
+  /**
+   * Wire uses the signed long type to store unsigned longs. Sigh. But when we encode as JSON we
+   * need to emit an unsigned value.
+   */
+  static final JsonAdapter<Long> UINT64_JSON_ADAPTER = new JsonAdapter<Long>() {
+    // 2^64, used to convert sint64 values >= 2^63 to unsigned decimal form
+    private final BigInteger power64 = new BigInteger("18446744073709551616");
+    private final BigInteger maxLong = BigInteger.valueOf(Long.MAX_VALUE);
+
+    @Override public Long fromJson(JsonReader reader) throws IOException {
+      BigInteger bigInteger = new BigInteger(reader.nextString());
+      return bigInteger.compareTo(maxLong) > 0
+          ? bigInteger.subtract(power64).longValue()
+          : bigInteger.longValue();
+    }
+
+    @Override public void toJson(JsonWriter writer, Long value) throws IOException {
+      if (value < 0) {
+        BigInteger unsigned = power64.add(BigInteger.valueOf(value));
+        writer.value(unsigned);
+      } else {
+        writer.value(value);
+      }
+    }
+  }.nullSafe();
+
+  /**
+   * Tragically Moshi doesn't know enough to follow a {@code @Uint64 List<Long>} really wants to be
+   * treated as a {@code List<@Uint64 Long>} and so we have to do it manually.
```

Comment:
https://github.com/square/moshi/issues/666

Reply:
even better than linking on the codereview thread is linking in the source code. That way if somebody is editing this later they can see if that bug is closed and this code can be simplified!